sympy/matrices/random.py

```diff
@@ -0,0 +1,2467 @@
+    [           sqrt(2)/4, sqrt(2)*(-1 - 2*I)/4, sqrt(2)*(-1 - I)/4],
+    [                 I/2,                  I/2,         -1/2 - I/2]])
+
+    >>> u = unitary(3, scalars=(-1,))
+    >>> simplify(u)
+    Matrix([
+    [-1,  0,  0],
+    [ 0, -1,  0],
+    [ 0,  0, -1]])
+
+    >>> s = sqrt(2)/2
+    >>> z = simplify(s + s * I)
+    >>> z
+    sqrt(2)*(1 + I)/2
+    >>> abs(z)
+    1
+    >>> spec = [-1, z, 1]
+    >>> unitary(3, spec=spec, length=0)  # no more random
+    Matrix([
+    [-1,                       0, 0],
+    [ 0, sqrt(2)/2 + sqrt(2)*I/2, 0],
+    [ 0,                       0, 1]])
+
+    >>> u = unitary(3, spec=spec)
+    >>> simplify(u)
+    Matrix([
+    [1/4 - I/4 + sqrt(2)*(2 + I)/4,  0, 1/4 - I/4 + sqrt(2)*I/4],
+    [                            0, -1,                       0],
+    [     -1/4 - sqrt(2)*I/4 + I/4,  0, 3/4 + I/4 + sqrt(2)*I/4]])
+    >>> simplify(u * u.H)
+    Matrix([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]])
+
+    >>> simplify(u.det())
+    sqrt(2)*(-1 - I)/2
+
+    >>> ev = u.eigenvals(simplify=True, multiple=True)
+    >>> ev = sorted(ev, key=(lambda x: re(x.evalf())))
+    >>> ev
+    [-1, sqrt(2)*I*(1 - I)/2, 1]
+    >>> # check if spec and ev coincide
+    >>> not any(simplify(x-y) for x, y in zip(spec, ev))
+    True
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    spec : tuple or list of symbols (optional),
+        set of values of which scalars (diagonal entries)
+        of :func:`isometry_normal` matrix.
+
+        Each list item has to be
+
+        * either 1 or -1
+        * or any complex unit $z$ (complex number with $|z|=1$)
+        * or a tuple of cosine value $c$ and sine value $s$
+          of a rotation square (the $z = c + s * I$)
+        * or just cosine value $c$ (then, a corresponding
+          sin value $s$ will be drawn randomly).
+
+    scalars : tuple or list of symbols (optional),
+        default values for random choosen scalar of
+        rotations to build the orthogonal matrix
+        see :func:`orthogonal`
+    length : integer (optional with default 2 * **dim**)
+        number rotations to build the matrix.
+        see :func:`orthogonal`
+    seed : random seed (optional)
+        see :func:`orthogonal`
+
+    See Also
+    ========
+    isometry_normal
+    orthogonal
+
+    """
+    rng = _rand(seed)
+    if spec is None:
+        if length == 0:
+            return _eye(dim)
+        length = length or 2 * dim
+
+        half = int(length / 2)
+        u = orthogonal(dim, scalars=scalars, length=half, seed=rng)
+        d = isometry_normal(dim, scalars, real=False, seed=rng)
+        v = orthogonal(dim,
+                       scalars=scalars, length=length - half, seed=rng)
+        return _multiply(u, d, v)
+
+    normal_form = isometry_normal(dim, spec, real=False, seed=rng)
+    s = unitary(dim, scalars=scalars, length=length, seed=rng)
+    return _multiply(_c(s), normal_form, s)
+
+
+def normal(dim,
```

Comment:
When I test this function it always seems to return a symmetric matrix. I would have thought that symmetric matrices could be made more easily than this though.

Reply:
Yes, but only if all _spec_ entries are real, this gives a symmetric matrix `M`. But this one is different from simply some `A*A.T` since `M` has a given set of eigenvalues (as set by _spec_). Note, those ev must all be unit roots.

If any _spec_ entry is complex, this yields a complex normal matrix `M`, i.e. `M.H*M=M*M.H` (which is true for symmetric matrices). Again, with a given set of eigenvalues.

Reply:
Maybe the docstring could be clearer about what this expected to return

Reply:
I tried to improve the docs on _unitary_ and _normal_ matrices.